doc: Add known issue for TX processor block (#20282)
nrfconnect/sdk-nrf: doc/nrf/releases_and_maturity/known_issues.rst

Bluetooth LE
============

> .. rst-class:: v2-9-0-nRF54H20-rc1 v2-9-0 v2-8-0

Reviewer: ```suggestion
.. rst-class:: v2-9-0-nRF54H20-1 v2-9-0 v2-8-0
```
Overlooked this on the first round.

Reviewer 2: Since "v2-9-0-nRF54H20-1" is not yet released and the release date is not yet known, we could keep this as -rc1, or other option could be that I can add this KI in my PR - https://github.com/nrfconnect/sdk-nrf/pull/19745, where the v2-9-0-nRF54H20-1 tag is added and this PR could be closed.
If this PR is merged now, we will have the v2.9.0-nrf54h20-1 tag in the public doc.

Author: Changed it back :)

> NCSDK-31528: Deadlock on sysworkq with tx_notify in Host
>   If a blocking callback is run on the system workqueue when the HCI buffer pool is full, the TX processor will be blocked.
>   This will prevent TX buffers from being freed, deadlocking the application.
>   An example of a blocking callback is calling the :c:func:`bt_conn_get_tx_power_level` function in a receive callback, which is blocking since it uses the :c:func:`bt_hci_cmd_send_sync` function.

Reviewer: ```suggestion
NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
  If a blocking callback is run on the system workqueue when the HCI buffer pool is full, the TX processor will be blocked.
  This prevents TX buffers from being freed, deadlocking the application.
  An example of a blocking callback is calling the :c:func:`bt_conn_get_tx_power_level` function in a receive callback, which is blocking since it uses the :c:func:`bt_hci_cmd_send_sync` function.
```

  **Workaround:** Do not use blocking calls in the system workqueue.

>   Alternatively, increase the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI buffer count.

Reviewer: ```suggestion
  Alternatively, increase the value of the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI buffer.
```

> .. rst-class:: v2-9-0-nRF54H20-1 v2-9-0 v2-8-0
> 
> NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
>   If a blocking callback is run on the system workqueue when the HCI buffer pool is full, the TX processor will be blocked.
>   This prevents TX buffers from being freed, deadlocking the application.
>   An example of a blocking callback is calling the :c:func:`bt_conn_get_tx_power_level` function in a receive callback, which is blocking since it uses the :c:func:`bt_hci_cmd_send_sync` function.
> 
>   **Workaround:** Do not use blocking calls in the system workqueue.
>   Alternatively, increase the value of the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI buffer.
>   This does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.

Reviewer 3: My two cents to improve clarity:

```suggestion
NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
  If a blocking callback (which in-turn needs to trigger an HCI command to complete its task) is run on the system workqueue when the HCI command buffer pool is full, the TX processing code will get blocked as well.
  This situation prevents TX command buffers from being freed, deadlocking the application.
  An example of a blocking callback is calling the :c:func:`bt_conn_get_tx_power_level` function in a receive callback. Calling such function can result in deadlock, since it uses the :c:func:`bt_hci_cmd_send_sync` function to complete its operation. The operation won't complete if HCI command buffer is full resulting in deadlock.

  **Workaround:** Do not use blocking calls in tasks triggered on the system workqueue.
  Alternatively, increase the value of the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI buffer.
  This does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.
```

Reviewer: Leave out "situation". Remember, all sentences on their own lines.

Reviewer: Would it be any better if we leave out "resulting in deadlock"? I think that is the redundant part.

Author: Think so, thanks!

Reviewer 3: I would retain that as an explainer for the previous paragraph. The example explains why deadlock happens.

Reviewer 3: The last two sentences are explaining how deadlock occurs in the context of the chosen example. It is easy for us to see it through, but it may not be easy for a typical reader of this text. I suggest to not remove too much for the sake of clarity.

>   The latter does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.

Reviewer: ```suggestion
  This does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.
```

Bluetooth LE
============

.. rst-class:: v2-9-0-nRF54H20-rc1 v2-9-0 v2-8-0

NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
  If a blocking callback (which in turn needs to trigger an HCI command to complete its task) is run on the system workqueue when the HCI command buffer pool is full, the TX processing code will get blocked.

>   This prevents TX command buffers from being freed, deadlocking the application.

Reviewer 3: ```suggestion
  This situation prevents TX command buffers from being freed, deadlocking the application.
```

Otherwise it is not clear what does "this" refers to.

Author: IMO it's obviously referring to the statement/line above, and adding "situation" doesn't really contribute (to the situation :)).

> NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
>   If a blocking callback (which in turn needs to trigger an HCI command to complete its task) is run on the system workqueue when the HCI command buffer pool is full, the TX processing code will get blocked.
>   This prevents TX command buffers from being freed, deadlocking the application.
>   An example of a blocking callback is calling the :c:func:`bt_conn_get_tx_power_level` function in a receive callback.
>   Calling such a function can result in a deadlock, since it uses the :c:func:`bt_hci_cmd_send_sync` function to complete its operation.
>   The operation won't complete if the HCI command buffer is full.
> 
>   **Workaround:** Do not use blocking calls in tasks triggered on the system workqueue.
>   Alternatively, increase the value of the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI buffer.
>   This does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.

Reviewer 4: There are several points here that are IMO not clear for customers as they are internal details of the Bluetooth Host implementation. I'd suggest to rephrase this to something like this:

```suggestion
  Blocking the system workqueue can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.
  The Bluetooth Host uses the system workqueue to complete processing of transmitted ACL data.
  Thus, using any blocking API on the system workqueue blocks the Bluetooth Host.
  If the HCI commands pool is empty, a blocking Bluetooth Host API that uses the :c:func:`bt_hci_cmd_send_sync` function is called from any thread (including the system workqueue) and the system workqueue is blocked, the deadlock occurs.

  **Workaround:** Do not block the system workqueue.
  Alternatively, use :kconfig:option:`CONFIG_BT_CONN_TX_NOTIFY_WQ`.
  You can also increase the value of the :kconfig:option:`CONFIG_BT_BUF_CMD_TX_COUNT` Kconfig option to increase the HCI commands pool.
  The latter does not guarantee that the problem is solved, as multiple blocking calls may exhaust the buffer pool.
```

Bluetooth LE
============

.. rst-class:: v2-9-0-nRF54H20-rc1 v2-9-0 v2-8-0

> NCSDK-31528: Deadlock on sysworkq with ``tx_notify`` in Host
>   Blocking the system workqueue if the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.

Reviewer 4: I think `tx_notify` is not something that customers are familiar with.
```suggestion
NCSDK-31528: Blocking the system workqueue if the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.
```

  The Bluetooth Host uses the system workqueue to complete processing of transmitted ACL data.
  Thus, using any blocking API on the system workqueue blocks the Bluetooth Host.

>   If the HCI commands pool is empty, and a blocking Bluetooth Host API that uses the :c:func:`bt_hci_cmd_send_sync` function is called from any thread (including the system workqueue) and the system workqueue is blocked, the deadlock can occur.

Reviewer 4: ```suggestion
  If the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled, and the HCI commands pool is empty, and a blocking Bluetooth Host API that uses the :c:func:`bt_hci_cmd_send_sync` function is called from any thread (including the system workqueue), and the system workqueue is blocked, the deadlock can occur.
```

Author: I added instead some bullet points to try to give a better view of this, please have a look!

Bluetooth LE
============

.. rst-class:: v2-9-0-nRF54H20-rc1 v2-9-0 v2-8-0

> NCSDK-31528: Blocking the system workqueue if the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.

Reviewer 3: Sentence parsing error. Something went wrong in the rephrasing and now it combination of two sentences with two verbs into single one.

Perhaps?
```suggestion
NCSDK-31528: Issue with blocking of the system workqueue. If the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.
```

Reviewer: The first line with the Jira issue ID is usually the same as the title in the issue, without a fullstop. Therefore, suggesting:
NCSDK-31528: Deadlock on system workqueue with `tx_notify` in host
  If the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled, blocking of the system workqueue can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.

Reviewer: Second line must be indented a couple of spaces (does not seem to show it properly in my comment.

Bluetooth LE
============

.. rst-class:: v2-9-0-nRF54H20-rc1 v2-9-0 v2-8-0

NCSDK-31528: Deadlock on system workqueue with ``tx_notify`` in host
  If the :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled, blocking of the system workqueue can cause a deadlock in the Bluetooth Host when running out of buffers in the HCI commands pool.
  The Bluetooth Host uses the system workqueue to complete processing of transmitted ACL data.
  Thus, using any blocking API on the system workqueue blocks the Bluetooth Host.
  For the deadlock to occur, the following must happen simultaneously:

>     * The :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled.
>     * The system workqueue is blocked.
>     * The HCI commands pool is empty.
>     * A blocking Bluetooth Host API that uses the :c:func:`bt_hci_cmd_send_sync` function is called from any thread (including the system workqueue).

Reviewer: ```suggestion
  * The :kconfig:option:`CONFIG_BT_HCI_ACL_FLOW_CONTROL` Kconfig option is disabled.
  * The system workqueue is blocked.
  * The HCI commands pool is empty.
  * A blocking Bluetooth Host API that uses the :c:func:`bt_hci_cmd_send_sync` function is called from any thread (including the system workqueue).
```
